File: cpp/kmp.cpp
// Subject to the terms of the Mozilla Public License, v. 2.0, see https://mozilla.org/MPL/2.0/.
// Free for non-commercial use, contact owner for commercial use.
//
// KMP algo.
// problem:
//   given a patterns P, len = n
//   also a text string T, len = m, get all the matches of P in T
// main steps:
// 1. build the auxilliary array lps[n] to store the jump index for current failed/ended match
// 2. do the search with that lps.
// So total time is O(n+m)

#include <iostream>
#include <vector>
using namespace std;

// build the lps array from pattern, time: O(n)
// lps[i] = x means for pattern P[0:n-1]
// P[0:x-1] == P[i+1-x:i]
// thus x holds the longest proper prefix that's also the suffix for substring P[0:i] (len = i+1)
int* build_lps(const string& p)
{
    int n = p.size();
    int* lps = new int[n];

    lps[0] = 0; // lps[0] is always 0
  
    // the iteration set lps[i] for i = 1:n-1
    for (int i = 0; i < n; i++)
    {
        // best case is this
        lps[i+1] = lps[i] + 1;
        // this is checked by if p[x] ?= p[i+1]
        // if yes, go on, 
        // if not, then next best case is lps[i+1] = lps[x] + 1 and checked by p[lps[i+1] -1] ?= p[i+1], again...
        // till lps[i+1] is 0
        // here lps[i+1] -1 is the 'x'
        while (lps[i+1] && p[lps[i+1] -1] != p[i+1])
        {
            // next best case
            int next = lps[lps[i+1] -1] + 1;
            if (lps[i+1] == next)
                // no more chance
                lps[i+1] = 0;
            else
                lps[i+1] = next;
        }
    }

    cout << "lps: \t";
    for (int i = 0; i < n; i++)
    {
        cout << "[" << i << "]: " << lps[i] << ", ";
    }
    cout << endl; 
    return lps;
}

// do search, time: O(m)
void kmp_search(const int* lps, const string& t, const string& p, vector<int>& result)
{
    int n = p.size();
    for (int i = 0, j = 0; i < t.size(); i++)
    {
        cout << "current states: i :" << i << " j: " << j << endl;
        // if keeping found, go on
        if (t[i] == p[j])
        {
            j++;
            // one match found, store the result, reset j 
            if (j == n) 
            {
                result.push_back(i);
                j = lps[n-1];
            }
        }
        // if not, and j not 0, jump to next lps
        else if (j)
        {
            // it's current j to match, so we need to find lps for 'before j': P[0:j-1] (len = j)
            j = lps[j-1];
            // check current i with new j again
            i--;
        }
        // else already checked P0, current i cannot match, just move on
    }
}

int main()
{
    string t = "ADAABAACAADAABAABA";
    string p =  "AABA";

    cout << "text:\t\t" << t << endl;
    cout << "pattern:\t" << p << endl;
    int* lps = build_lps(p);


    vector<int> result;
    kmp_search(lps, t, p, result);


    cout << "pattern matches at: \t";
    for (const auto& pos : result)
    {
        cout << pos << ", ";
    }
    cout << endl;

    delete [] lps;
    return 0;
}

File: gopher/src/rozen/kmp/kmp.go
// Subject to the terms of the Mozilla Public License, v. 2.0, see https://mozilla.org/MPL/2.0/.
// Free for non-commercial use, contact owner for commercial use.
//
// KMP algo. demo
// Problem: given a pattern P, len=n, and a text T, len=m, find the matches of P in T.
// KMP way: the key is for the cases of 'retry new match, either after a complete one or failed one',
//          we don't need to start from P[0], instead, based on our knowledge of P,
//          we can jump to start from P[x>0] directly,
//          the jump index of each index in P is in an array called lps[n]
//          for the proving, see blog.
// So the total time will be just: O(n+m)
package main
import "fmt"

//  build lps, iteration way, time:O(n)
func build_lps(P string) []int {
    n := len(P)
    lps := make([]int, n)
    // init of iteration, jump index of 0 is of cause 0.
    lps[0] = 0

    for i := 0; i < n - 1; i++ {
        // use ptr for clearer code
        ptr := &lps[i + 1]
        // current best case, let lps[i] = x
        *ptr = lps[i] + 1
        // case true? determined by: p[x] ?= p[i+1]
        for *ptr > 0 && P[*ptr - 1] != P[i + 1] {
            // next best case lps[x]+1
            next := lps[*ptr - 1] + 1
            if *ptr > next {
                *ptr = next
            } else {
                // no hope, has to be p[0]
                *ptr = 0
            }
        }
    }
    return lps
}

// do search with lps, just linear, so: O(m)
func kmp_search(T, P string, lps []int, result *[]int) {
    m := len(T)
    n := len(P)

    // i for T, j for P
    for i, j := 0, 0; i < m; i++ {
        // if matches go on
        if T[i] == P[j] {
            j++
            // found pattern, reset j
            if j == n {
                *result = append(*result, i)
                j = lps[n - 1]
            }
        } else if j > 0 {
            // not matches, reset j
            // check the blog, the trick here is lps[j-1] is to get the jump index
            // for string P[0:j-1] (len=j), thus enable the next matching of P[x] vs T[i]
            j = lps[j - 1]
            // revist i
            i--
        }
        // else i cannot match any
    }
}

func main() {
    T := "ABAABAACAADAABAABA"
    P := "AABA"
    result := new([]int)

    lps := build_lps(P)
    kmp_search(T, P, lps, result)

    fmt.Printf("text:\t")
    fmt.Println(T)
    fmt.Printf("pattern:\t")
    fmt.Println(P)
    fmt.Printf("lps:\t")
    fmt.Println(lps)
    fmt.Printf("result:\t")
    fmt.Println(*result)
}
